### PR TITLE
feat(api): async workflow deletion

### PR DESCRIPTION
**Pull request by fsamin**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/workflow_run.go [hunk 1]
```diff
@@ -687,16 +686,11 @@ func (api *API) postWorkflowRunHandler() Handler {
...
-					runtime.Stack(buf, true)
-					log.Error("[PANIC] workflow.PurgeWorkflowRun> %s", string(buf))
-				}
-			}()
-			workflow.PurgeWorkflowRun(api.mustDB(), *wf)
-		}()
+		sdk.GoRoutine(
+			"workflow.PurgeWorkflowRun",
+			func() {
+				workflow.PurgeWorkflowRun(api.mustDB(), *wf)
```
Comment (golangci) [inline engine/api/workflow_run.go]: Error return value of `workflow.PurgeWorkflowRun` is not checked

File: engine/api/workflow/dao.go [hunk 2]
```diff
@@ -629,11 +634,21 @@ func Update(db gorp.SqlExecutor, store cache.Store, w *sdk.Workflow, oldWorkflow
 	return updateLastModified(db, store, w, u)
 }
 
+// MarkAsDelete marks a workflow to be deleted
+func MarkAsDelete(db gorp.SqlExecutor, w *sdk.Workflow) error {
+	if _, err := db.Exec("update workflow set to_delete = true where id = $1", w.ID); err != nil {
+		return sdk.WrapError(err, "Delete> Unable to mark as delete workflow id %d", w.ID)
```
Comment (bnjjj) [inline engine/api/workflow/dao.go]: s/Delete/MarkAsDelete

**Comment by fsamin** [comment]
you're assuming, that the user is able to delete a generated workflow (with a very few number of runs) and generated a new workflow faster than the goroutine is running at the end of `deleteWorkflowHandler` ?

i don't think it's possible and i think it's not useful to cover this nearly impossible usecase, especially because the 'migration' will be dropped in a few weeks.

**Comment by yesnault** [comment]
@fsamin ok. The migrate use delete and not markAsDelete 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: engine/api/purge/purge.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+	for _, w := range wrkflws {
+		tx, err := db.Begin()
+		if err != nil {
+			return sdk.WrapError(err, "purge.Workflows> unable to start tx")
+		}
+
+		proj, has := projects[w.ProjectID]
+		if !has {
+			continue
```
Comment (sguiheux) [inline engine/api/purge/purge.go]: _ = tx.Rollback()

File: engine/api/purge/purge.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+		if err := workflow.Delete(tx, store, &proj, &w); err != nil {
+			log.Error("purge.Workflows> unable to delete workflow %d: %v", w.ID, err)
+			_ = tx.Rollback()
+			continue
+		}
+
+		if err := tx.Commit(); err != nil {
+			log.Error("purge.Workflows> unable to commit tx: %v", err)
+			continue
```
Comment (sguiheux) [inline engine/api/purge/purge.go]: _ = tx.Rollback()

**Pull request merged**